asllib/ASTUtils.ml

```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
 
 let is_right_increasing = function
   | MUL | SHL | SHR | POW | PLUS -> true
-  | DIV | DIVRM | MINUS -> false
-  | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MINUS -> false
+  | DIV | DIVRM | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_decreasing = function
-  | DIVRM | DIV | MINUS ->
+  | MINUS ->
       true
       (* DIV is here because its right-hand-side is always meant to be positive *)
   | MUL | SHL | SHR | POW | PLUS -> false
-  | MOD | AND | BAND | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | DIVRM | DIV | MOD | AND | BAND | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let () =
-  if false then (
-    let all_binops_here = [ DIVRM; DIV; MINUS; MUL; SHL; SHR; POW; PLUS ] in
+  if true then (
+    let all_binops_here = [ MINUS; MUL; SHL; SHR; POW; PLUS ] in
     List.iter
       (fun b -> assert (is_right_decreasing b || is_right_increasing b))
       all_binops_here;
     List.iter (fun b -> assert (is_left_increasing b)) all_binops_here)
 
 (* Begin ConstraintBinop *)
 let constraint_binop =
+  let exact e = Constraint_Exact e and range a b = Constraint_Range (a, b) in
   let constraint_binop_pair op c1 c2 =
     match (c1, c2) with
-    | Constraint_Exact e1, Constraint_Exact e2 ->
-        Constraint_Exact (binop op e1 e2)
+    | Constraint_Exact e1, Constraint_Exact e2 -> exact (binop op e1 e2)
     | Constraint_Exact e1, Constraint_Range (e21, e22) ->
-        if is_right_increasing op then
-          Constraint_Range (binop op e1 e21, binop op e1 e22)
-        else Constraint_Range (binop op e1 e22, binop op e1 e21)
+        if is_right_increasing op then range (binop op e1 e21) (binop op e1 e22)
+        else range (binop op e1 e22) (binop op e1 e21)
     | Constraint_Range (e11, e12), Constraint_Exact e2 ->
-        Constraint_Range (binop op e11 e2, binop op e12 e2)
+        range (binop op e11 e2) (binop op e12 e2)
     | Constraint_Range (e11, e12), Constraint_Range (e21, e22) ->
         if is_right_increasing op then
-          Constraint_Range (binop op e11 e21, binop op e12 e22)
-        else Constraint_Range (binop op e11 e22, binop op e12 e21)
+          range (binop op e11 e21) (binop op e12 e22)
+        else range (binop op e11 e22) (binop op e12 e21)
   in
   let constraint_mod = function
-    | Constraint_Exact e | Constraint_Range (_, e) ->
-        Constraint_Range (zero_expr, e)
+    | Constraint_Exact e | Constraint_Range (_, e) -> range zero_expr e
+  in
+  let constraint_divisions op c1 c2 =
+    match (c1, c2) with
+    | Constraint_Exact e1, Constraint_Exact e2 -> [ exact (binop op e1 e2) ]
+    | Constraint_Range (e11, e12), Constraint_Exact e2 ->
+        [ range (binop op e11 e2) (binop op e12 e2) ]
+    | Constraint_Exact e1, Constraint_Range (_e21, e22) ->
```

Comment:
Is this correct?
For example `10 DIV 1..10` = `1, 2, 5, 10`, not `[10..1, 1..10]`